File: js/src/lib/runner.js
import path from "node:path"
import { readFile, writeFile } from "node:fs/promises"
import { spawn } from 'node:child_process'
import { inspect } from "node:util"

global.inspect = (obj) => {
  console.log(inspect(obj, {depth: null, colors: true}))
}

const COPY_BIN = "pbcopy"
const DATA_DIR = "data"
const inputFilePath = (input) => path.join(DATA_DIR, `${input}.input`)
const outputFilePath = (input, part) => path.join(DATA_DIR, `${input}_${part}.output`)

const red = "\u{001B}[31m"
const green = "\u{001B}[32m"
const bold = "\u{001B}[1m"
const reset = "\u{001B}[0m"
const checkmark = "✓"
const cross = "✗"

// testOutput: (testKey: string, solveFunc: (input: string) -> string)
export const testOutput = async (inputFile, part, solveFunc) => {
  const [inputData, outputData] = await Promise.all([
    readFile(inputFilePath(inputFile)),
    readFile(outputFilePath(inputFile, part)),
  ])

  const expected = outputData.toString().trim()
  const input = inputData.toString().trim()
  const result = String(solveFunc(input)).trim()

  if (result === expected) {
    console.log(`${green}${bold}${checkmark} Actual output matches expected output for data '${inputFile}' part ${part}${reset}`)
  } else {
    console.error(`${red}${bold}${cross} Actual output did not match expected output for data '${inputFile}' part ${part}${reset}`)
    console.log("Expected:")
    console.log(expected)
    console.log("Got:")
    console.log(result)
  }
}

export const printOutput = async (inputFile, solveFunc) => {
  const inputData = await readFile(inputFilePath(inputFile))
  const input = inputData.toString().trim()
  const result = String(solveFunc(input)).trim()
  console.log(`Result for input ${inputFile}:`)
  console.log(result)
}

export const writeOutput = async (inputFile, part, solveFunc) => {
  const inputData = await readFile(inputFilePath(inputFile))
  const input = inputData.toString().trim()
  const result = String(solveFunc(input)).trim()
  await writeFile(outputFilePath(inputFile, part), result)
  console.log(`Writing result for input ${inputFile} part ${part} to file`)
}

// This currently only works on macOS
export const copyOutput = async (inputFile, solveFunc) => {
  const inputData = await readFile(inputFilePath(inputFile))
  const input = inputData.toString().trim()
  const result = String(solveFunc(input)).trim()

  return new Promise((resolve, reject) => {
    const copyProcess = spawn(COPY_BIN)

    copyProcess.on('exit', (code) => {
      if (code === 0) {
        console.log(`The output for key ${inputFile} has been copied to the clipboard`)
        resolve()
      } else {
        reject(new Error(`Error copying output for ${inputFile}`))
      }
    })

    copyProcess.stdin.end(result)
  })
}

File: cpp/src/lib/runner.cpp
#include "runner.h"
#include <algorithm>
#include <cctype>
#include <fstream>
#include <iostream>
#include <stdexcept>

static const std::string dataDir = "data";
static const std::string red = "\033[31m";
static const std::string green = "\033[32m";
static const std::string bold = "\033[1m";
static const std::string reset = "\033[0m";
static const std::string checkmark = "✓";
static const std::string cross = "✗";

// Trim from start (in place)
static inline void ltrim(std::string &s) {
    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) { return !std::isspace(ch); }));
}

// Trim from end (in place)
static inline void rtrim(std::string &s) {
    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) { return !std::isspace(ch); }).base(), s.end());
}

// Trim from both ends (in place)
static inline void trim(std::string &s) {
    ltrim(s);
    rtrim(s);
}

static std::string readFile(const std::string &relativePath) {
    std::ifstream file(relativePath);
    if (!file.is_open()) {
        throw std::runtime_error("Unable to read file: " + relativePath);
    }

    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    return content;
}

void testAocOutput(const std::string &dayKey, const std::string &inputName, const std::string &partKey,
                   const AocPartSolver &partSolver) {
    std::string inputFilePath = dataDir + "/day" + dayKey + "/" + inputName + ".input";
    std::string outputFilePath = dataDir + "/day" + dayKey + "/" + inputName + "_" + partKey + ".output";

    try {
        std::string input = readFile(inputFilePath);
        trim(input);

        std::string expectedOutput = readFile(outputFilePath);
        trim(expectedOutput);

        std::string result = partSolver(input);
        trim(result);

        if (result == expectedOutput) {
            std::cout << bold << green << checkmark << " SUCCESS: Day " << dayKey << " part " << partKey << " "
                      << inputName << ". Result matches expected output! (" << result << ")" << reset << std::endl;
        } else {
            std::cout << bold << red << cross << " FAIL: Day " << dayKey << " part " << partKey << " " << inputName
                      << ". Expected " << expectedOutput << " but got " << result << reset << std::endl;
        }
    } catch (const std::exception &e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }
}
